delivery.go

```diff
@@ -1,70 +1,122 @@
 package rmq
 
 import (
+	"context"
 	"fmt"
+	"time"
 )
 
 type Delivery interface {
 	Payload() string
-	Ack() bool
-	Reject() bool
-	Push() bool
+
+	Ack() error
+	Reject() error
+	Push() error
 }
 
-type wrapDelivery struct {
+type redisDelivery struct {
+	ctx         context.Context
 	payload     string
 	unackedKey  string
 	rejectedKey string
 	pushKey     string
 	redisClient RedisClient
+	errChan     chan<- error
 }
 
-func newDelivery(payload, unackedKey, rejectedKey, pushKey string, redisClient RedisClient) *wrapDelivery {
-	return &wrapDelivery{
+func newDelivery(
+	ctx context.Context,
+	payload string,
+	unackedKey string,
+	rejectedKey string,
+	pushKey string,
+	redisClient RedisClient,
+	errChan chan<- error,
+) *redisDelivery {
+	return &redisDelivery{
+		ctx:         ctx,
 		payload:     payload,
 		unackedKey:  unackedKey,
 		rejectedKey: rejectedKey,
 		pushKey:     pushKey,
 		redisClient: redisClient,
+		errChan:     errChan,
 	}
 }
 
-func (delivery *wrapDelivery) String() string {
+func (delivery *redisDelivery) String() string {
 	return fmt.Sprintf("[%s %s]", delivery.payload, delivery.unackedKey)
 }
 
-func (delivery *wrapDelivery) Payload() string {
+func (delivery *redisDelivery) Payload() string {
 	return delivery.payload
 }
 
-func (delivery *wrapDelivery) Ack() bool {
-	// debug(fmt.Sprintf("delivery ack %s", delivery)) // COMMENTOUT
+// blocking versions of the functions below with the following behavior:
+// 1. return immediately if the operation succeeded or failed with ErrorNotFound
+// 2. in case of other redis errors, send them to the errors chan and retry after a sleep
+// 3. if redis errors occur after StopConsuming() has been called, ErrorConsumingStopped will be returned
+
+func (delivery *redisDelivery) Ack() error {
+	errorCount := 0
+	for {
+		count, err := delivery.redisClient.LRem(delivery.unackedKey, 1, delivery.payload)
+		if err == nil { // no redis error
+			if count == 0 {
+				return ErrorNotFound
+			}
+			return nil
+		}
+
+		// redis error
+
+		errorCount++
+
+		select { // try to add error to channel, but don't block
+		case delivery.errChan <- &DeliveryError{Delivery: delivery, RedisErr: err, Count: errorCount}:
```

Comment:
My reasoning for why I don't think this would be an issue:
1. Each goroutine only produces up to one error per second.
2. There's one goroutine for the heartbeat and one goroutine per consumer.
3. So even if you have thousands of consumers, you should have no trouble consuming all of them from the error channel in time. 
4. Plus: the channel gets created by the user and injected into rmq, so they could make it really big if needed.

Reply:
i c, makes sense